### PR TITLE
Keep comments in annotations when converting them to attributes

### PR DESCRIPTION
**Pull request by carlos-granados**
Sometimes annotations may include some comments after the annotation definition. For example:

```php
    /**
     * @Route("/path", name="action") api route
     */
```

The current rector rule that converts annotations to attributes ignores these comments which are lost. This PR modifies the code so that these comments are preserved as line comments added after the attribute like this:

```php
    #[Route(path: '/path', name: 'action')] // api route
```

See the comments in the PR for more info

### CODE REVIEW

File: rules/CodeQuality/Rector/Class_/CompleteDynamicPropertiesRector.php [hunk 1]
```diff
@@ -154,7 +154,9 @@ private function shouldSkipClass(Class_ $class): bool
             return true;
         }
 
-        return $class->extends instanceof FullyQualified && ! $this->reflectionProvider->hasClass($class->extends->toString());
+        return $class->extends instanceof FullyQualified && ! $this->reflectionProvider->hasClass(
```
Comment (carlos-granados) [inline rules/CodeQuality/Rector/Class_/CompleteDynamicPropertiesRector.php:157]: When I ran `composer complete-check` it found several unrelated code standard fixes. I have included them in the PR so that all tests pass even if they are unrelated to the current code change

File: rules/Php80/Rector/Class_/AnnotationToAttributeRector.php [hunk 2]
```diff
@@ -76,7 +76,7 @@ public function getRuleDefinition(): RuleDefinition
 class SymfonyRoute
 {
     /**
-     * @Route("/path", name="action")
+     * @Route("/path", name="action") api route
```
Comment (carlos-granados) [inline rules/Php80/Rector/Class_/AnnotationToAttributeRector.php:79]: Includes a sample of this functionality in the documentation

File: src/BetterPhpDocParser/PhpDoc/DoctrineAnnotationTagValueNode.php [hunk 3]
```diff
@@ -18,11 +19,16 @@ public function __construct(
...
         array $values = [],
-        ?string $silentKey = null
+        ?string $silentKey = null,
+        ?string $comment = null
     ) {
         $this->hasChanged = true;
 
         parent::__construct($values, $originalContent, $silentKey);
+
+        if ($comment) {
```
Comment (carlos-granados) [inline src/BetterPhpDocParser/PhpDoc/DoctrineAnnotationTagValueNode.php]: When creating the tag value node, if the tag had a comment, we save it as an attribute in the node to be used later

File: src/BetterPhpDocParser/PhpDocParser/DoctrineAnnotationDecorator.php [hunk 4]
```diff
@@ -409,14 +409,20 @@ private function createDoctrineSpacelessPhpDocTagNode(
             $currentPhpNode
         );
 
+        $comment = $this->staticDoctrineAnnotationParser->getCommentFromRestOfAnnotation(
```
Comment (carlos-granados) [inline src/BetterPhpDocParser/PhpDocParser/DoctrineAnnotationDecorator.php:412]: Once we have processed the key/value pairs for the annotation we read the rest of the annotation as a comment

File: src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php [hunk 5]
```diff
@@ -21,6 +21,12 @@
  */
 final readonly class StaticDoctrineAnnotationParser
 {
+    /**
+     * @var string
+     * @see https://regex101.com/r/Pthg5d/1
+     */
+    private const END_OF_VALUE_CHARACTERS_REGEX = '/^[)} \n"\']+$/i';
```
Comment (carlos-granados) [inline src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php]: This is used to go past the end of the values for the annotation as the token iterator is not advanced to the end of these values

File: src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php [hunk 6]
```diff
@@ -80,6 +86,24 @@ public function resolveAnnotationValue(
...
+        // we skip all the remaining tokens from the end of the declaration of values
+        while (
+            preg_match(self::END_OF_VALUE_CHARACTERS_REGEX, $tokenIterator->currentTokenValue())
+        ) {
+            $tokenIterator->next();
+        }
+
+        // the remaining of the annotation content is the comment
+        $comment = substr($annotationContent, $tokenIterator->currentTokenOffset());
+        // we only keep the first line as this will be added as a line comment at the end of the attribute
```
Comment (carlos-granados) [inline src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php:109]: This solution is not ideal as some info may be lost but could not think of a better alternative, as adding all lines in a single line seemed wrong and adding several line comments seemed also wrong.

File: src/NodeTypeResolver/Node/AttributeKey.php [hunk 7]
```diff
@@ -301,4 +301,9 @@ final class AttributeKey
      * @var string
      */
     public const IS_USED_AS_ARG_BY_REF_VALUE = 'is_used_as_arg_by_ref_value';
+
+    /**
+     * @var string
+     */
+    public const ATTRIBUTE_COMMENT = 'attribute_comment';
```
Comment (carlos-granados) [inline src/NodeTypeResolver/Node/AttributeKey.php:308]: This constant is used to save the comments in the nodes

File: src/PhpAttribute/NodeFactory/PhpAttributeGroupFactory.php [hunk 8]
```diff
@@ -87,7 +87,12 @@ public function create(
         $attributeName->setAttribute(AttributeKey::PHP_ATTRIBUTE_NAME, $annotationToAttribute->getAttributeClass());
 
         $attribute = new Attribute($attributeName, $args);
-        return new AttributeGroup([$attribute]);
+        $attributeGroup = new AttributeGroup([$attribute]);
+        $comment = $doctrineAnnotationTagValueNode->getAttribute(AttributeKey::ATTRIBUTE_COMMENT);
```
Comment (carlos-granados) [inline src/PhpAttribute/NodeFactory/PhpAttributeGroupFactory.php:91]: If the tag value node had a comment we copy it to the attribute group node

File: src/PhpParser/Printer/BetterStandardPrinter.php [hunk 9]
```diff
@@ -139,6 +139,16 @@ protected function p(Node $node, $parentFormatPreserved = false): string
             : $content;
     }
 
+    protected function pAttributeGroup(Node\AttributeGroup $node): string
```
Comment (carlos-granados) [inline src/PhpParser/Printer/BetterStandardPrinter.php:142]: When printing an attribute, if we had saved a comment, we add it as a line comment at the end of the attribute

File: src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php [hunk 10]
```diff
@@ -80,6 +86,24 @@ public function resolveAnnotationValue(
...
+        $commentLines = explode("\n", $comment);
```
Comment (samsonasik) [inline src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php]: windows new line need to be supported, you can do `Strings::split()` like in 

https://github.com/rectorphp/rector-src/blob/a4445df500e51deb893319e180c3fbdf64e8dc25/src/Console/Formatter/ColorConsoleDiffFormatter.php#L62

The `\r?\n` is actually more reliable imo

Comment (samsonasik) [reply src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php]: For reference for `\r?\n` usage https://github.com/search?q=repo%3Arectorphp%2Frector-src%20%5Cr%3F%5Cn&type=code

File: src/BetterPhpDocParser/PhpDoc/DoctrineAnnotationTagValueNode.php (hunk 3 above)
Comment (samsonasik) [reply src/BetterPhpDocParser/PhpDoc/DoctrineAnnotationTagValueNode.php]: this should be strict whether "0" or empty string "" or null, I guess you can do:

```php
if (! in_array($comment, ["", null], true))
```

since "0" comment is valid

File: src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php (hunk 10 above)
Comment (carlos-granados) [reply src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php]: Updated

File: src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php [hunk 11]
```diff
@@ -21,6 +22,18 @@
  */
 final readonly class StaticDoctrineAnnotationParser
 {
+    /**
+     * @var string
+     * @see https://regex101.com/r/qduj2O/1
+     */
+    private const NEWLINES_REGEX = "#\n\r|\n#";
```
Comment (samsonasik) [inline src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php]: could you update to `\r?\n` as it seems more reliable?

File: src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php (hunk 5 above)
Comment (samsonasik) [reply src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php]: windows new line need to be supported as well

```
'/^[)} \r?\n"\']+$/i';

File: src/PhpParser/Printer/BetterStandardPrinter.php [hunk 12]
```diff
@@ -139,6 +139,16 @@ protected function p(Node $node, $parentFormatPreserved = false): string
...
+    {
+        $ret = '#[' . $this->pCommaSeparated($node->attrs) . ']';
```
Comment (samsonasik) [inline src/PhpParser/Printer/BetterStandardPrinter.php]: original `parent::pAttributeGroup()` should be called first imo

File: src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php (hunk 11 above)
Comment (carlos-granados) [reply src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php]: done

File: rules/Php80/Rector/Class_/AnnotationToAttributeRector.php (hunk 2 above)
Comment (samsonasik) [reply rules/Php80/Rector/Class_/AnnotationToAttributeRector.php:79]: I think separate fixture is better, and add next docblock example for `/** ` usage

File: src/PhpParser/Printer/BetterStandardPrinter.php (hunk 12 above)
Comment (carlos-granados) [reply src/PhpParser/Printer/BetterStandardPrinter.php]: Good shout, updated

File: src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php (hunk 5 above)
Comment (carlos-granados) [reply src/BetterPhpDocParser/PhpDocParser/StaticDoctrineAnnotationParser.php]: We don't need the ? as this just checks for any of these characters, so adding \r is enough, updated

File: rules/Php80/Rector/Class_/AnnotationToAttributeRector.php (hunk 2 above)
Comment (samsonasik) [reply rules/Php80/Rector/Class_/AnnotationToAttributeRector.php:79]: multi line comment fixture seems also needed

Comment (samsonasik) [reply rules/Php80/Rector/Class_/AnnotationToAttributeRector.php:79]: Oh, i see, it is example config

**Review by samsonasik — APPROVED** [review]

**Comment by samsonasik** [comment]
Thank you @carlos-granados , let's give it a try 👍
